### PR TITLE
CSV export fails for table with composite primary key

### PR DESCRIPTION
**Pull request by adrinr**
## Description
Support composite primary keys on filters, allowing actions such as CSV exporting specific rows.

## Addresses
- [BUDI-7065 - CSV export fails for table with composite primary key](https://linear.app/budibase/issue/BUDI-7065/csv-export-fails-for-table-with-composite-primary-key)

## Screenshoots
_Table with name+age as primary key_

https://github.com/Budibase/budibase/assets/15987277/e8879d5f-4912-4c5b-950d-eba64f323d71

image

## Launchcontrol
Allow exporting CSV exports for tables with composite primary key

### CODE REVIEW

File: packages/server/src/api/routes/tests/row.spec.ts [hunk 1]
```diff
@@ -1452,6 +1436,46 @@ describe.each([
         { status: 404 }
       )
     })
+
+    // MSSQL needs a setting called IDENTITY_INSERT to be set to ON to allow writing
+    // to identity columns. This is not something Budibase does currently.
+    providerType !== DatabaseName.SQL_SERVER &&
+      it("should handle filtering by composite primary keys", async () => {
+        const tableRequest = saveTableRequest({
+          primary: ["number", "string"],
```
Comment (samwho) [inline packages/server/src/api/routes/tests/row.spec.ts:1445]: The last time I was looking into this, for external data sources, only the first element in this array gets used. Are you sure that this test is actually creating a composite primary key?

Comment (adrinr) [reply packages/server/src/api/routes/tests/row.spec.ts:1445]: Not 100% sure if in the database it gets created as identity, but for the logic of these export tests we shoudn't really care. We want to treat both columns as the primary key, and we want to get sure that the select queries work as expected (ignoring if they are or are not primary keys on the external dbs).

Even so, we could double-check it, but I think it gets actually set on the DB. For this reason, I needed to disable SQL Server, as we were getting actual DB exceptions

Comment (adrinr) [reply packages/server/src/api/routes/tests/row.spec.ts:1445]: Just to clarify, without this PR code changes, this test would fail

**Review by samwho — APPROVED** [review]

File: packages/backend-core/src/sql/sql.ts [hunk 2]
```diff
@@ -170,6 +170,8 @@ function convertBooleans(query: SqlQuery | SqlQuery[]): SqlQuery | SqlQuery[] {
   return query
 }
 
+const COMPLEX_ID_OPERATOR = "_complexIdOperator"
```
Comment (mike12345567) [inline packages/backend-core/src/sql/sql.ts:173]: Could we move this perhaps into type (I know its not strictly a type, but we put enums and other stuff in there as well, we probably would want to use it for typing in the future as well

Comment (adrinr) [reply packages/backend-core/src/sql/sql.ts:173]: You mean in @budibase/types? We could also add strict types if they live in backend core, as this is where the only "usage" lives

Comment (adrinr) [reply packages/backend-core/src/sql/sql.ts:173]: Amended!

**Review by mike12345567 — APPROVED** [review]
Very nice typing!

**Pull request merged**